rclcpp_lifecycle/include/rclcpp_lifecycle/managed_entity.hpp

```diff
@@ -0,0 +1,46 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_LIFECYCLE__MANAGED_ENTITY_HPP_
+#define RCLCPP_LIFECYCLE__MANAGED_ENTITY_HPP_
+
+#include <atomic>
+
+namespace rclcpp_lifecycle
+{
+
+///  Base class for lifecycle entities, like `LifecyclePublisher`.
+class ManagedEntityInterface
```

Comment:
The interface in rclpy is a bit different (see [here](https://github.com/ros2/rclpy/blob/38095981409f000c733b4c53037aa31ab0a76279/rclpy/rclpy/lifecycle/managed_entity.py#L23-L47)):

- on_activate()/on_deactivate() receives the previous state as a parameter and can fail.
  - I think in most of the cases on_activate()/on_deactivate() shouldn't fail, the on_shutdown()/on_configure() transitions should be doing the fallible work.
  - Receiving the previous state is not important for this transitions, as it's always 'inactive' for `on_activate()` and 'activated'  for `on_deactivate()`.
- The rclpy interface allows you to also provide `on_shutdown()`/`on_configure()`/`on_error()` transitions.
  - LifecyclePublisher is not using it, though entities could lazely create resources on `on_configure()` and destroy them on `on_shutdown()` if those methods were available.
  - The default on_configure/on_shutdown() implementation in Node might be less useful, as it's not clear what to do if some of the callbacks fail (see [this](https://github.com/ros2/rclpy/blob/38095981409f000c733b4c53037aa31ab0a76279/rclpy/rclpy/lifecycle/node.py#L188-L189) and [this](https://github.com/ros2/rclpy/blob/38095981409f000c733b4c53037aa31ab0a76279/rclpy/rclpy/lifecycle/node.py#L197-L198)).
    In practice, I think people usually want to manually write on_configure()/on_shutdown().

We could either modify this PR to do exactly what rclpy does or reduce rclpy flexibility to match more closely what this PR is adding.
I prefer a bit this less flexible version, which is what we actually seem to need.

Reply:
> I think in most of the cases on_activate()/on_deactivate() shouldn't fail

Having less failure points sounds nice to me. I haven't thought through scenarios where someone may want to fail during activation/deactivation, but we could add it if the need arises. 

> The default on_configure/on_shutdown() implementation in Node might be less useful, as it's not clear what to do if some of the callbacks fail

If one or more of the callbacks fail, I would think we would want to transition all entities to the `unconfigued` state (as indicated by the design doc diagram). What do you think?

I agree that in most cases people will probably provide their own on_configure/on_shutdown logic. I think that it still might be nice if we call the transitions on all managed entities though. E.g. call any user-defined callback for the node and then, if that is successful, call the transition callbacks for managed entities. Or do you see something wrong with that behavior?

Reply:
> If one or more of the callbacks fail, I would think we would want to transition all entities to the unconfigued state (as indicated by the design doc diagram). What do you think?

I think we could do that.
The logic would be:
- configure
  - Try to transition all managed entities to configured
  - If transition fails (TRANSITION_FAILURE), shut down all the managed entities that already transitioned to go back to the unconfigured state.
  - If any callback errors (TRANSITION_ERROR), then we immediately return error without doing any more transitions.
    This applies when a on_configure callback errors, or when an on_shutdown callback errors after a previous failure.
- shutdown
  - Similar, but on failures we do nothing, and continue trying to move the rest of the entities to the shutdown state.

The only problem I see is that the logic is a bit complex and might not be actually useful in practice (we don't have any lifecycle entity that do work on configure/shutdown currently).

> I agree that in most cases people will probably provide their own on_configure/on_shutdown logic. I think that it still might be nice if we call the transitions on all managed entities though. E.g. call any user-defined callback for the node and then, if that is successful, call the transition callbacks for managed entities. Or do you see something wrong with that behavior?

I think this has been resolved in the discussion [here](https://github.com/ros2/demos/pull/552#discussion_r784217616).

Reply:
Fair enough. I'm fine keeping the implementation simple for now; we can add complexity if the need arises.